opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/AbstractDaoHibernate.java

```diff
@@ -259,8 +269,42 @@ public void deleteAll(final Collection<T> entities) throws DataAccessException {
 
     @Override
     public List<T> findMatching(final org.opennms.core.criteria.Criteria criteria) {
-        final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
-        return getHibernateTemplate().execute(callback);
+        if(criteria.isMultipleAnd()){
+            return findMultiAndMatching(criteria);
+        } else {
+            final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
+            return getHibernateTemplate().execute(callback);
+        }
+    }
+
+    private List<T> findMultiAndMatching(final org.opennms.core.criteria.Criteria criteria){
+        Set<T> allUniqueRecords = new LinkedHashSet<>();
+        Collection<Restriction> allRestrictions = criteria.getRestrictions();
+
+        //set of multiand restrictions
+        Set<Restriction> multiAndRestirctionSet = allRestrictions.stream().filter(
+                restriction -> restriction.getType().equals(Restriction.RestrictionType.MULTIAND)).collect(Collectors.toSet());
+
+        //set of non multiand restrictions
+        Set<Restriction>   nonMultiAndRestirctionSet =  Sets.difference(new HashSet<Restriction>(allRestrictions),multiAndRestirctionSet);
+
+        //iterating multiand and setting nonmultiand + single multiand restriction in criteria,
+        // retrieve result set, create intersection of each query executed
+        multiAndRestirctionSet.stream().forEach(restriction ->{
+            Collection<Restriction> allMultiAndRestrictions = ((AllRestriction) restriction).getRestrictions();
+            allMultiAndRestrictions.stream().forEach(singleMultiAndRestriction ->{
+                criteria.setRestrictions(nonMultiAndRestirctionSet);
+                criteria.addRestriction(singleMultiAndRestriction);
```

Comment:
This modifies the `Criteria`. Is it reusable afterwards?

Reply:
another copy of criteria create and used for query execution, original criteria will be inact